src/tests/shuttle.md

```diff
@@ -0,0 +1,205 @@
+
+Testing concurrent code is hard because there is often an unbounded number of possible executions. There are multiple
+approaches to it: [exhaustive space exploration](https://docs.rs/loom/latest/loom/) and [stochastic simulation](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/asplos277-pct.pdf). Shuttle takes the latter approach and uses randomized
+scheduling techniques to detect concurrency issues with high probability. Certain IPA layers benefit more from having
+those tests
+
+#### Infrastructure
+Infrastructure uses tasks and schedulers to keep up with the load, so it benefits greatly from having good coverage
+of concurrency tests. In fact, [this issue](https://github.com/private-attribution/ipa/issues/256) was revealed only
+after onboarding infrastructure to Shuttle.
+
+Every module that uses `Arc`, `thread`, `tokio::spawn` benefits from concurrency testing.
+
+#### Protocols
+Protocols are inherently asynchronous and vulnerable to the same class of concurrency issues as the infrastructure.
+
+```rust
+async fn sum_of_products(a: &[u32], b: &[u32]) -> u64 {
+    let mut sum: u64 = 0;
+    for x in a {
+        for y in b {
+            let r = secure_mul(x, y).await;
+            sum += r;
+        }
+    }
+    
+    sum
+}
+```
+
+
+### How does it work?
+
+There are multiple strategies to run a Shuttle test, most common one used in IPA involves randomized concurrency testing.
+Shuttle picks a possible execution based on random choice. There is more in Shuttle than just a randomized scheduler, more information can be found
+[here](https://docs.rs/shuttle/latest/shuttle/scheduler/index.html).
+
+### Getting started
+
+Shuttle tests live inside the `tests/` folder. The first line of each test module should be the following
+
+```rust
+#![cfg(all(feature = "shuttle", test))]
+```
+
+Shuttle test is a regular Rust test that uses Shuttle scheduler to run it. It cannot be made async, so `block_on` is used
+to block the main thread awaiting future completion.
+
+```rust
+#[test]
+fn my_first_concurrency_test() {
+    // how many times Shuttle needs to schedule this test for execution.
+    // Each execution will have a unique schedule, so keep it as high as possible.
+    let iterations = 1000; 
+    shuttle::check_random(|| { shuttle::future::block_on(async {
+        // run semi-honest multiplication
+        let world = TestWorld::new().await;
+
+        let a = Fp31::from(2);
+        let b = Fp31::from(5);
+
+        let res = world.semi_honest((a, b), |ctx, (a, b)| async move { 
+            ctx.multiply(RecordId::from(0), &a, &b).await.unwrap() 
+        }).await;
+
+        // assertions are important because some corruptions lead
+        // to incorrect results rather than panic or deadlock
+        assert_eq!(a * b, res.reconstruct());
+    }) }, iterations);
+}
+```
+
+### Running Shuttle tests
+
+Execute the cargo `test` command with shuttle feature enabled:
+```bash
+cargo test --features shuttle
```

Comment:
You might want to point out that this generally disables most other tests, which are async tests that won't run with the shuttle scheduler.